File: udemy-masterclass/Sorting/07radixSort.js
function getDigit(num, place) {
  return Math.floor(Math.abs(num) / Math.pow(10, place)) % 10
}

function digitCount(num) {
  if (num === 0) return 1;
  return Math.floor(Math.log10(Math.abs(num))) + 1;
}

function mostDigits(nums) {
  let maxDigits = 0;
  for (let num of nums) {
    maxDigits = Math.max(maxDigits, digitCount(num));
  }
  return maxDigits;
}


function radixSort(nums) {
  for (let i = 0; i < mostDigits(nums); i++) {
    let buckets = Array.from({length: 10}, () => []);

    for (let num of nums) {
      let digit = getDigit(num, i);
      buckets[digit].push(num);
    }
    nums = [].concat(...buckets)

  }

  return nums;
}

console.log(radixSort([4, 1, 2, 3, 9, 10]))

File: udemy-masterclass/Sorting/07radix_sort.py
import math
import itertools

def get_digit_at(num, position):
    return int(math.floor(abs(num) / math.pow(10, position)) % 10)

def get_digit_count(num):
    if num == 0: return 1
    return int(math.floor(math.log10(abs(num))) + 1)

def max_digit(arr):
    maximum = 0
    for num in arr:
        if get_digit_count(num) > maximum:
            maximum = get_digit_count(num)
    return maximum


def radix_sort(arr):
    for i in range(max_digit(arr)):
        buckets = [[] for _ in range(10)]
        for num in arr:
            buckets[get_digit_at(num, i)].append(num)
        arr = list(itertools.chain(*buckets))
    return arr

print(radix_sort([4, 3, 2, 4, 111]))
